### PR TITLE
Basic diff

### PR DESCRIPTION
**Pull request by thetheodor**
Missing:
- [x] Parallelism (only the generation is done in parallel)
- [x] Documentation
- [x]  Testing
- [x] Merging #58

### CODE REVIEW

**Comment by thetheodor** [comment]
@Quarub this is the most basic diff testing functionality

It's still incomplete so feel free to ignore this.

**Comment by thetheodor** [comment]
@Quarub This implements a very basic version of the differential testing part. Can be used, e.g., as:
`python -m dead "clang-14 -O3" "clang-15 -O3"    output_directory --testing_mode=unidirectional` 
This will generate and test in parallel cases and dump (in a very stupid way) the results in output_directory

File: dead/differential_testing.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    """
+
+    program: InstrumentedProgram
+    setting1: CompilationSetting
+    setting2: CompilationSetting
+    markers_only_eliminated_by_setting1: tuple[DCEMarker | VRMarker, ...]
+    markers_only_eliminated_by_setting2: tuple[DCEMarker | VRMarker, ...]
+
+
+class DifferentialTestingMode(Enum):
```
Comment (thetheodor) [inline dead/differential_testing.py:55]: I don't really like the naming scheme, but I couldn't come up with something better. I'm not sure if this mode is needed in the long run.

**Comment by thetheodor** [comment]
I'll add reductions in a subsequent PR

**Review by Quarub — DISMISSED** [review]
I'm wondering if dead2 should follow the "data hoarding" approach like I used in the ORM version, i.e. instead of storing the difference of markers, we store all alive markers and have the interestingness test as a separate function.

Pro: Nearly free signatures and signature tests. Test-type can be decided after the fact.

Con: Complexity.

**Comment by thetheodor** [comment]
> I'm wondering if dead2 should follow the "data hoarding" approach like I used in the ORM version, i.e. instead of storing the difference of markers, we store all alive markers and have the interestingness test as a separate function.
> 
> Pro: Nearly free signatures and signature tests. Test-type can be decided after the fact.
> 
> Con: Complexity.

Yup, it probably makes sense to directly store the marker status (probably other stuff as well?), but atm it's very easy to extract them: `case.program.find_{alive,dead}_markers(case.setting{1,2})`. 

I am wondering if it would make sense to directly store some kind of signature instead of plainly listing markers

**Pull request merged**